vlm/src/main/scala/geotrellis/contrib/vlm/RasterSource.scala

```diff
@@ -43,7 +43,12 @@ import geotrellis.util.GetComponent
   * @groupprio reproject 2
   */
 trait RasterSource extends CellGrid with AutoCloseable with Serializable {
+  // sfitch: As with `targetCellType` below, this feels like API leakage, and seems to be captured already by
+  // the `XXXResampleRasterSource` classes.
   def resampleMethod: Option[ResampleMethod]
```

Comment:
@metasim do you think its ever valuable to know if you're working with original or resample/reprojected data? If so relying on `XXXResampledRasterSource` would require a `match` on all possible instances of it. If not, then yeah, this is leaky.

Reply:
@echeipesh From a purists perspective, if you do  then the abstraction is wrong. It's an implementation detail that shouldn't affect how you treat the abstraction called "RasterSource". But in essence, it's The Expression Problem, and abstracting in one direction requires specificity in another. On balance, I'd rather move that specificity away from the public API into the internal implementation.

Reply:
This actually leads me to another point which is RasterSource currently mixes (at least) two concerns that may deserve to be separated:

1. How to read (sub-)rasters (efficiently)
2. How to do that while efficiently transforming the data into the desired form without large data copies

(Do you agree? Have I missed any?)

The first is obvious and core to the abstraction. The second is technically a performance optimization, but a required one. My gut says this responsibility could be separately abstracted, using something akin to an AST (maybe even a linear one), providing easier testing, reusability, and less implementation bleeding.

Reply:
Something like this: